File: python/位运算/260. 只出现一次的数字 III.py
from typing import List
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        mask = 0
        for i in nums:
            mask ^= i
        diff = mask & (-mask)
        res = 0
        for i in nums:
            # 在该位不同，以这一位区分出两个数字，这里最终必然会只剩下一个数字
            if i & diff:
                res ^= i
        return [res,res^mask]

File: go/位运算/260. 只出现一次的数字 III.go
package main

func singleNumber(nums []int) []int {
	mask := 0
	for _, v := range nums {
		mask ^= v
	}
	diff := mask & (-mask)

	res := 0
	for _, v := range nums {
		if diff&v != 0 {
			res ^= v
		}
	}
	return []int{res, res ^ mask}
}
